src/Microsoft.AspNetCore.OData/Formatter/ODataOutputFormatter.cs

```diff
@@ -299,4 +292,53 @@ private MediaTypeHeaderValue GetContentType(string contentTypeValue)
             return contentType;
         }
     }
+
+    internal class StreamWrapper : Stream
```

Comment:
What's this for?

Reply:
The WebAPI OData library ends up calling a lower level OData Library with an internal JSON reader/writer.  That reader/writer currently calls synchronous Read/Write methods.  So, by wrapping the body.Stream in this StreamWrapper before passing it to the lower level library we can make sure that the synchronous APIs are never called.

Would love your thoughts here, but seemed an easy/lightweight way to make sure that no synchronous method was called (if the lower level library already calls the asynch methods it's just pass-through).

---
In reply to: [515376424](https://github.com/OData/WebApi/pull/2336#discussion_r515376424) [](ancestors = 515376424)

Reply:
This is a good way to verify that but I would add it to tests, not to the main project.

Reply:
Unfortunately, it's not just a test verification issue.  We *know* that, deep within the lower level OData Library that this library calls, the code that actually reads/writes to the stream currently calls synchronous read/write methods. It will take a while to fix that lower level library to be async all of the way to the wire, so this makes sure that, at the lowest level, even though the underlying OData Library makes a synchronous call to read/write, we are always calling the asynchronous APIs on the stream.

Are we likely to run into issues with this approach? The main idea of this overall fix is to not set the "AllowSynchronousIO=true" flag, which unblocks *other* asynchronous calls.  Is there a better way to handle this until we are able to rev the lower level library?

---
In reply to: [515953270](https://github.com/OData/WebApi/pull/2336#discussion_r515953270) [](ancestors = 515953270)